opponent.h

```diff
@@ -8,11 +9,14 @@ namespace pong
 class Opponent : public GameObject
 {
 private:
-    float mSpeed { 0.0f };
+    glm::vec3 mVelocity { 0.0f };
+    std::shared_ptr<Ball> mBall { nullptr };
```

Comment:
This seems incorrect. Does the opponent OWN the ball? I think it likely just has a passing interest in it. I would say make this a plain pointer or plain ref.

Reply:
Yeah I guess that makes more sense. Although is either pointer or ref better? I feel I have heard of "no raw pointers" which instictually makes me use smart pointers instead. It seems like this type of situation complicates it though.

Reply:
Yep, you're close to the right saying! 

The actual saying is "no raw new" meaning you should never see just the use of `new` in a codebase, indicating they should use make_unique or similar to do the same thing. (There is also "no raw loops" which means don't use a for-loop if there is a std::algorithm that works, like `std::any_of` that we used in an earlier review).


In general, here is how you should think about the 3 major pointer types. They are all fundamentally in terms of "ownership". When you use a pointer, think about where in your abstraction the thing is OWNED.

Unique pointer:
- One unique owner
- If you hold the unique pointer, you are that objects owner

Shared pointer:
- Multiple owners
- If you hold a shared pointer, you are one of multiple owners
- too many owners of a shared pointer is a code smell

Raw pointer (or reference):
- Non-owning reference to an object
- If you hold a raw pointer, or reference, you are likely just borrowing it. You have no ownership responsibilities

Sources:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-owner
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-ptr
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#r4-a-raw-reference-a-t-is-non-owning